### ISSUE TITLE
Wrong JSON Format Export

### ISSUE THREAD

**Issue by RolandSall** [nicolargo/glances#1814]
#### Description

I am writing an application that monitors the resources variation of the system. I need to get a JSON file and later on use the data to plot a graph. I am using the following command

`glances --export json --export-json-file ./"my directory" --disable-plugin percpu,memswap,network,diskio,fs,docker,sensors,ip,uptime`

I am exporting the JSON with only the data that matters to me.

The problem is that the exported JSON has a wrong format; The JSON is missing the "[  ]" and the commas that separates the extracted objects.

Example:
```
--> This "[" is missing (added manually) 
[
  {
    "cpu": {
      "total": 25.7,
      "user": 22.9,
      "system": 2.5,
             .... 
    },
    "load": {
      "min1": 1.3,
      "min5": 2.12,
          .....
    },
    "mem": {
      "total": 33639563264,
          .....
    },
    "network": {},
    "fs": {},
    "processcount": {
      "total": 314,
           ....
    },
    "system": {
      "os_name": "Linux",
              ....
    },
    "sensors": {},
    "gpu": {
      "0.key": "gpu_id",
           ....
    }
  },   The "," is also added manually
  {
    "cpu": {
      "total": 25.8,
      "user": 16.4,
      "system": 2.0,
           ....
    },
    "load": {
      "min1": 1.28,
      "min5": 2.1,
          ....
    },
    "mem": {
      "total": 33639563264,
      "available": 27626160128,
                ...
    },
    "network": {},
    "fs": {},
    "processcount": {
      "total": 314,
           ....
    },
    "system": {
      "os_name": "Linux",
                 ....
    },
    "sensors": {},
    "gpu": {
      "0.key": "gpu_id",
      "0.gpu_id": 0,
           ....
    }
  }
]
```

Glances v3.1.6.2 with PsUtil v5.8.0.

Is there any other way to extract JSON data with the right format?

The format needed is the following:

```
[
{"First Data Fetch"},
{"Second Data Fetch"},
{"Third Data Fetch"},
.....
{"N Data Fetch"}
]
```

**Comment by Hierosme** [comment]
Hello.

For me start the data structure by a "[]" have nothing relatif to the JSON specification, from the documentation that is a object.

and normaly nothing on JSON format impose to put the Data Structure on a Object [].

The seconde strange thing on you exemple is the float type. I remenber hard time about type ;)

I have cross it type of trouble many time (Like Python2 vs Python3 code).

But for  me you exemple is more Python data structure like 

```
import json
json.dumps(the_thing[0])
```

Can you explain more about why you apply haven't any trouble and how the workarround happen ?

**Comment by RolandSall** [comment]
I could not quite well understand your question. Let me rephrase what is happening;
The command that applied above from the command line prompt (regardless if am using any specific language).:

`glances --export json --export-json-file ./"my directory" --disable-plugin percpu,memswap,network,diskio,fs,docker,sensors,ip,uptime` 

Will generate supposedly a JSON file. However, if you use any online JSON formatter (i.e: https://jsonformatter.curiousconcept.com/) it will reject it due to an invalid format.

This is why I pointed about that if it possible to include "[]" and the commas to validate its format

**Comment by Hierosme** [comment]
In can reproduce a regression with the version **Glances v3.1.6.2 with PsUtil v5.8.0** (from pip) and **Glances v3.1.0 with psutil v5.5.1** (Debian 10 via ap)

**Glances v3.1.6.2 with PsUtil v5.8.0**
```
/home/USER/.local/bin/glances --export json --export-json-file ./here.json --disable-plugin percpu,memswap,network,diskio,fs,docker,sensors,ip,uptime
```
produse a invalide JSON format

```
Error: Parse error on line 99:
...		"steal": 0.0	}} {	"load": {		"min1
---------------------^
Expecting 'EOF', '}', ',', ']', got '{'
```

**Comment by Hierosme** [comment]
it look like a valid an well formated JSON string it have been append on the output file where it should replace existing value.
i see that because each line is a update of value's.

Then we should search arround output file JSON where a the open functionis call with the append flag :)

**Comment by RolandSall** [comment]
Yes, the first time this command run, it will generate a valide JSON object. However, once another iteration occurs, the second object is not append it to the first one. It is updated on a second line which makes the total JSON file with invalid syntax.

What I want since this command above will run indefinitely, I want once I terminated to obtain a Valid JSON format file to observe the variation of my resources while running certain applications.

**Comment by Hierosme** [comment]
i think i got the trouble

the file : https://github.com/nicolargo/glances/blob/develop/glances/exports/glances_json.py make the trouble.

the __init__ function make the open with "w" flag ()
```
        # Set the JSON output file
        try:
            if PY3:
                self.json_file = open(self.json_filename, 'w')
            else:
                self.json_file = open(self.json_filename, 'wb')
        except IOError as e:
            logger.critical("Cannot create the JSON file: {}".format(e))
            sys.exit(2)
```
and that the export function just write the data from self.buffer

```
            # Export stats to JSON file
            data_json = json.dumps(self.buffer)
            self.json_file.write("{}\n".format(data_json))
```

Ok cool but where is the trouble ?

* Nothing close the file descriptor, that finaly the export function make a write in a allready open file (then no replace happen)
* The __init__ is a good place for the raise Error about Json, but here __enter__ is a best candidate
* I'll personally remake a open inside the export function, with a **with**

For me the init should be
```
        # Set the JSON output file
        try:
            if PY3:
                self.json_file = open(self.json_filename, 'w')
                self.json_file.close()
            else:
                self.json_file = open(self.json_filename, 'wb')
                self.json_file.close()
        except IOError as e:
            logger.critical("Cannot create the JSON file: {}".format(e))
            sys.exit(2)
```

And for the export function
```
    # Export stats to JSON file
    if PY3:
        with open('self.json_filename', "w") as file_object:
            file_object.write("{}\n".format(json.dumps(self.buffer)))
    else:
        with open('self.json_filename', "wb") as file_object:
            file_object.write("{}\n".format(json.dumps(self.buffer)))
            
```

**Comment by Hierosme** [comment]
motivation for tests and PR ?

**Comment by RolandSall** [comment]
Yes, I agree with where you pinned point the issue.  

> Nothing close the file descriptor, that finaly the export function make a write in a allready open file (then no replace happen)

This explain why the JSON objects are being added in an incorrect matter. 

Should this be issued as a bug to be fixed?

**Comment by Hierosme** [comment]
we have to create a PR and flag the issue #1814 in the commit message.

Let me propose a PR

**Comment by Hierosme** [comment]
@RolandSall could you test if teh fixe work for you ?

**Comment by RolandSall** [comment]
The same error still persists for me, Maybe I am setting up the environment in the wrong way. I will double-check. What steps did you follow ?

**Comment by Hierosme** [comment]
Sorry the clone if from my fork because my PR is not include yet.

I have test from my point of view it fixe the trouble

```
mkdir TMPDIR
cd TMPDIR
git clone https://github.com/Hierosme/glances.git
cd glances
python3 -mvenv venv
. venv/bin/activate
pip install --editable .
./venv/bin/glances --export json --export-json-file ./here.json --disable-plugin percpu,memswap,network,diskio,fs,docker,sensors,ip,uptime
```
CTRL + C
```
cat ./here.json
cd ../..
rm -rf ./TMPDIR
```

**Comment by RolandSall** [comment]
I tried above and in my case, only one object was added with no extra appended JSON objects. This might be due to my environment. I will leave this up to the creator to see his suggestion about it

**Comment by Hierosme** [comment]
Nop that me who as code the fact Only one entry write each time it have a iteration.

You reaquire a Valid Json incrementally ?

That is a supper bad idea due to the size it can take for open , insert , and rewrite the file.
It have no simple value for make it, exempte how it have been done (One valid JSON format by line), the output file is not JSON valid but you can store value (one json valid by line) with append method. It it have been code originally.

What that the Manual ? is it suppose to be one JSON valid file or one JSON entry by line ?

**Comment by Hierosme** [comment]
Hmmmm it look to have a methode with list().expend(json.load()), Let me fews minutes i'll make a try

**Comment by Hierosme** [comment]
Sorry i can't make it work, to mush difficult. Then it have tow way.

* One Valide JSON by line, but he file is not a True JSON file
* Only one entry by file, then you have only one entry but that a true json file

Personnaly without file rotation policy, i recommand the frist option, because from my point that should not be the monitor application to store the sensor memory. Then periodically somethin come search the JSon file

**Comment by RolandSall** [comment]
So it should be originally an array of JSON Objects (valid syntax format)

```
[
{"First Data Fetch"},
{"Second Data Fetch"},
{"Third Data Fetch"},
.....
{"N Data Fetch"}
]
```

I do not think it is a bad practice, because, in the end, the command is to extract a JSON file. What purpose would it serve if each object is added on a new line with an invalid syntax

**Comment by Hierosme** [comment]
I understand, you want store the memory of each **glances iteration** inside the json file.

for me it's really hard to code, you can make a try with a **pip install --editable ." and edit ``glances/exports/glances_json.py``, by start
``` shell
. venv/bin/activate
venv/bin/glances
```

You can make all you tests. You'll certainlly see the touble of want store memory inside a signle json file.

I have re test and a recusive Json Import to clean export look complicated to do

**Comment by Hierosme** [comment]
i continue to think the best way is just the fresh data of the actual iteration of glances. Then with curl you can send the json file to you Alert Manager.

Keep the memory of iteration without a TimeStamp is from my point a bad think. What is containt of the file should be the last fresh data ...
